Add scripts

joshmoore opened the pull request:
Utility scripts for working with LDAP servers.

cc: @pwalczysko @sbesson

sbesson committed:
Add initialization LDIF file

sbesson committed:
Add instructions for setting up the LDAP server

sbesson committed:
Port user script from apacheds-docker

sbesson committed:
Use real creds

sbesson committed:
Improve user/group logic

sbesson committed:
Add description of the user/group scripts

sbesson committed:
Clear tree

sbesson committed:
Add initialization script

sbesson committed:
Fix user/group scripts to use LDAPDOMAIN

sbesson committed:
Default to dc=openmicroscopy,dc=org domain

sbesson committed:
Fix Readme

sbesson committed:
Remove init.ldif

sbesson committed:
Add search script to return the default LDAP tree

sbesson committed:
Fix Readme

sbesson committed:
Add confirmation logic for clear script

joshmoore committed:
Migrate ldap-ci to minke

Rather than use DNS on beluga, ldap-ci is now the well-known
port `10389` on minke (production docker). Storage will still
be deleted on every restart.

joshmoore committed:
Cleanup scripts

joshmoore committed:
Run scripts in test

sbesson reviewed (approved):
Makes sense to me. Release as `0.3.0` given the addition of utility API?

joshmoore commented:
No objections. Note: these scripts need some API work which is why they weren't previously in this repository. Happy to adapt that work around @pwalczysko's deadlines.

pwalczysko commented:
@joshmoore added https://github.com/joshmoore/apacheds-docker/pull/1 - feel free to close or accept it or tell me to close or improve or whatever is the best in this situatio (or merge this first, then we can deal with https://github.com/joshmoore/apacheds-docker/pull/1 if preferred.

joshmoore committed:
ldapo: rewrite of bash scripts using python-ldap

Script which can be used to either generated LDIF
or to talk directly to a server. Several assumptions
are made similar to those in the ome.ldif file.

Multiple passwords are supported.

Addition of a dummy user for creating a group is *not*.

joshmoore committed:
Add package to travis

pwalczysko commented:
@joshmoore So in order to install python for ldap to run these scripts, I am doing

```
    - name: Install python-ldap
      become: yes
      pip:
        name:
        - "python-ldap"
        state: present
        virtualenv: "/home/ldap-venv"
        virtualenv_site_packages: yes
```

but this fails with

```
TASK [Install python-ldap] *****************************************************************************
fatal: [ome-training-outreach.openmicroscopy.org]: FAILED! => {"changed": false, "cmd": "/home/ldap-venv/bin/pip install python-ldap", "failed": true, "msg": "stdout: Downloading/unpacking python-ldap\n  Running setup.py egg_info for package python-ldap\n    \n    file Lib/ldap.py (for module ldap) not found\n    file Lib/ldap/controls.py (for module ldap.controls) not found\n    file Lib/ldap/extop.py (for module ldap.extop) not found\n    file Lib/ldap/ldapobject.py (for module ldap.ldapobject) not found\n    file Lib/ldap/schema.py (for module ldap.schema) not found\n    warning: no files found matching 'Makefile'\n    warning: no files found matching 'Modules/LICENSE'\nRequirement already satisfied (use --upgrade to upgrade): setuptools in /home/ldap-venv/lib/python2.7/site-packages (from python-ldap)\nInstalling collected packages: python-ldap\n  Running setup.py install for python-ldap\n    file Lib/ldap.py (for module ldap) not found\n    file Lib/ldap/controls.py (for module ldap.controls) not found\n    file Lib/ldap/extop.py (for module ldap.extop) not found\n    file Lib/ldap/ldapobject.py (for module ldap.ldapobject) not found\n    file Lib/ldap/schema.py (for module ldap.schema) not found\n    file Lib/ldap.py (for module ldap) not found\n    file Lib/ldap/controls.py (for module ldap.controls) not found\n    file Lib/ldap/extop.py (for module ldap.extop) not found\n    file Lib/ldap/ldapobject.py (for module ldap.ldapobject) not found\n    file Lib/ldap/schema.py (for module ldap.schema) not found\n    \n    file Lib/ldap.py (for module ldap) not found\n    file Lib/ldap/controls.py (for module ldap.controls) not found\n    file Lib/ldap/extop.py (for module ldap.extop) not found\n    file Lib/ldap/ldapobject.py (for module ldap.ldapobject) not found\n    file Lib/ldap/schema.py (for module ldap.schema) not found\n    warning: no files found matching 'Makefile'\n    warning: no files found matching 'Modules/LICENSE'\n    building '_ldap' extension\n    gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DHAVE_TLS -DHAVE_LIBLDAP_R -DHAVE_SASL -DLDAPMODULE_VERSION=2.5.2 -DLDAPMODULE_AUTHOR=python-ldap project -DLDAPMODULE_LICENSE=Python style -IModules -I/usr/include -I/usr/include/sasl -I/usr/local/include -I/usr/local/include/sasl -I/usr/include/python2.7 -c Modules/LDAPObject.c -o build/temp.linux-x86_64-2.7/Modules/LDAPObject.o\n    In file included from Modules/LDAPObject.c:8:0:\n    Modules/errors.h:7:18: fatal error: lber.h: No such file or directory\n     #include \"lber.h\"\n                      ^\n    compilation terminated.\n    error: command 'gcc' failed with exit status 1\n    Complete output from command /home/ldap-venv/bin/python -c \"import setuptools;__file__='/home/ldap-venv/build/python-ldap/setup.py';exec(compile(open(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))\" install --record /tmp/pip-BFvfX4-record/install-record.txt --single-version-externally-managed --install-headers /home/ldap-venv/include/site/python2.7:\n    running install\n\nrunning build\n\nrunning build_py\n\nfile Lib/ldap.py (for module ldap) not found\n\nfile Lib/ldap/controls.py (for module ldap.controls) not found\n\nfile Lib/ldap/extop.py (for module ldap.extop) not found\n\nfile Lib/ldap/ldapobject.py (for module ldap.ldapobject) not found\n\nfile Lib/ldap/schema.py (for module ldap.schema) not found\n\ncreating build\n\ncreating build/lib.linux-x86_64-2.7\n\ncopying Lib/ldapurl.py -> build/lib.linux-x86_64-2.7\n\ncopying Lib/ldif.py -> build/lib.linux-x86_64-2.7\n\ncopying Lib/slapdtest.py -> build/lib.linux-x86_64-2.7\n\ncreating build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/__init__.py -> build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/async.py -> build/lib.linux-x86_64-2.7/ldap\n\ncreating build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/__init__.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/deref.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/libldap.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/openldap.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/ppolicy.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/psearch.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/pwdpolicy.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/readentry.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/sessiontrack.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/simple.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/sss.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/controls/vlv.py -> build/lib.linux-x86_64-2.7/ldap/controls\n\ncopying Lib/ldap/cidict.py -> build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/dn.py -> build/lib.linux-x86_64-2.7/ldap\n\ncreating build/lib.linux-x86_64-2.7/ldap/extop\n\ncopying Lib/ldap/extop/__init__.py -> build/lib.linux-x86_64-2.7/ldap/extop\n\ncopying Lib/ldap/extop/dds.py -> build/lib.linux-x86_64-2.7/ldap/extop\n\ncopying Lib/ldap/filter.py -> build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/functions.py -> build/lib.linux-x86_64-2.7/ldap\n\ncreating build/lib.linux-x86_64-2.7/ldap/ldapobject\n\ncopying Lib/ldap/ldapobject/__init__.py -> build/lib.linux-x86_64-2.7/ldap/ldapobject\n\ncopying Lib/ldap/ldapobject/simple.py -> build/lib.linux-x86_64-2.7/ldap/ldapobject\n\ncopying Lib/ldap/ldapobject/reconnect.py -> build/lib.linux-x86_64-2.7/ldap/ldapobject\n\ncopying Lib/ldap/logger.py -> build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/modlist.py -> build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/pkginfo.py -> build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/resiter.py -> build/lib.linux-x86_64-2.7/ldap\n\ncopying Lib/ldap/sasl.py -> build/lib.linux-x86_64-2.7/ldap\n\ncreating build/lib.linux-x86_64-2.7/ldap/schema\n\ncopying Lib/ldap/schema/__init__.py -> build/lib.linux-x86_64-2.7/ldap/schema\n\ncopying Lib/ldap/schema/models.py -> build/lib.linux-x86_64-2.7/ldap/schema\n\ncopying Lib/ldap/schema/subentry.py -> build/lib.linux-x86_64-2.7/ldap/schema\n\ncopying Lib/ldap/schema/tokenizer.py -> build/lib.linux-x86_64-2.7/ldap/schema\n\ncopying Lib/ldap/syncrepl.py -> build/lib.linux-x86_64-2.7/ldap\n\nfile Lib/ldap.py (for module ldap) not found\n\nfile Lib/ldap/controls.py (for module ldap.controls) not found\n\nfile Lib/ldap/extop.py (for module ldap.extop) not found\n\nfile Lib/ldap/ldapobject.py (for module ldap.ldapobject) not found\n\nfile Lib/ldap/schema.py (for module ldap.schema) not found\n\nrunning egg_info\n\nwriting requirements to Lib/python_ldap.egg-info/requires.txt\n\nwriting Lib/python_ldap.egg-info/PKG-INFO\n\nwriting top-level names to Lib/python_ldap.egg-info/top_level.txt\n\nwriting dependency_links to Lib/python_ldap.egg-info/dependency_links.txt\n\nwarning: manifest_maker: standard file '-c' not found\n\n\n\nfile Lib/ldap.py (for module ldap) not found\n\nfile Lib/ldap/controls.py (for module ldap.controls) not found\n\nfile Lib/ldap/extop.py (for module ldap.extop) not found\n\nfile Lib/ldap/ldapobject.py (for module ldap.ldapobject) not found\n\nfile Lib/ldap/schema.py (for module ldap.schema) not found\n\nreading manifest file 'Lib/python_ldap.egg-info/SOURCES.txt'\n\nreading manifest template 'MANIFEST.in'\n\nwarning: no files found matching 'Makefile'\n\nwarning: no files found matching 'Modules/LICENSE'\n\nwriting manifest file 'Lib/python_ldap.egg-info/SOURCES.txt'\n\nrunning build_ext\n\nbuilding '_ldap' extension\n\ncreating build/temp.linux-x86_64-2.7\n\ncreating build/temp.linux-x86_64-2.7/Modules\n\ngcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DHAVE_TLS -DHAVE_LIBLDAP_R -DHAVE_SASL -DLDAPMODULE_VERSION=2.5.2 -DLDAPMODULE_AUTHOR=python-ldap project -DLDAPMODULE_LICENSE=Python style -IModules -I/usr/include -I/usr/include/sasl -I/usr/local/include -I/usr/local/include/sasl -I/usr/include/python2.7 -c Modules/LDAPObject.c -o build/temp.linux-x86_64-2.7/Modules/LDAPObject.o\n\nIn file included from Modules/LDAPObject.c:8:0:\n\nModules/errors.h:7:18: fatal error: lber.h: No such file or directory\n\n #include \"lber.h\"\n\n                  ^\n\ncompilation terminated.\n\nerror: command 'gcc' failed with exit status 1\n\n----------------------------------------\nCleaning up...\nCommand /home/ldap-venv/bin/python -c \"import setuptools;__file__='/home/ldap-venv/build/python-ldap/setup.py';exec(compile(open(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))\" install --record /tmp/pip-BFvfX4-record/install-record.txt --single-version-externally-managed --install-headers /home/ldap-venv/include/site/python2.7 failed with error code 1 in /home/ldap-venv/build/python-ldap\nStoring complete log in /root/.pip/pip.log\n"}
	to retry, use: --limit @/Users/pwalczysko/Work/ansible-examples-omero/public-user/playbook.retry

PLAY RECAP *********************************************************************************************
ome-training-outreach.openmicroscopy.org : ok=93   changed=3    unreachable=0    failed=1   

```

pwalczysko commented:
@manics suggested that I might be able to install the python-ldap via yum, but then I would be missing some juicy options, like let the users run their python scripts (against OMERO)

Edited: now correctly addressing @manics

joshmoore commented:
> I am doing...

In this PR:

```
+addons:
+  apt:
+    packages:
+    - python-ldap
```

uses `apt-get` to install. If the same can be done with `yum install` then that would work fine. 

> @simleo 

@manics ?

> I would be missing some juicy options, like let the users run their python scripts (against OMERO)

I'm unsure what this means.

pwalczysko commented on bin/ldapo:
```diff
@@ -0,0 +1,284 @@
+#!/usr/bin/env python
+
```
@joshmoore for this new script, it works very nicely. But I would appreciate more verbose output.

```
[pwalczysko@ome-training-outreach ldap]$ ./ldapo --apply user public --password public
(105, [])
```
is too little information.
Maybe if I knew how to double check the users already in ldap ...

it could be even behind a ``--verbose`` flag

pwalczysko replied:
yes, actually the ``./search`` script is giving me the list. But still, it would be much more satisfying to have some informative output from the ``./ldapo`` as well

joshmoore replied:
I was surprised at how terse the output from `python-ldap` is. One option would be to run the related search and output that. Happy to add it (either with or without the `--verbose` flag).

pwalczysko replied:
I think so, without the ``--verbose`` is fine by me.

pwalczysko commented:
Re: https://github.com/openmicroscopy/apacheds-docker/pull/7#issuecomment-363842002 - I think I have managed to install the ``python-ldap`` as it is showing in ``pip freeze`` output and also your ./ldapo script now works nicely.

```
    - name: Prerequisites for ldap
      become: yes
      yum:
        name: "{{ item }}"
        state: present
      with_items:
        - openldap-clients
        - python-virtualenv
        - gcc
        - python-ldap
```

pwalczysko commented:
Now, I ``smuggled`` the ``./ldapo`` script to the outreach server using wget.
Would be a good plan to clone this repo onto the outreach server via ansible ?

joshmoore committed:
ldapo: add `get` command and print on update

`get ${dn}` now prints the record for the given entry.
After all updates, this method is called showing the
new status of the record.

joshmoore commented:
> I think so, without the --verbose is fine by me.

Updated.

> Would be a good plan to clone this repo onto the outreach server via ansible ?

It's certainly fine. Ansible can also call wget for you.

joshmoore commented:
Ach, noch eine Sache: better suggestions for the name than `ldapo`?

pwalczysko commented:
>Ach, noch eine Sache: better suggestions for the name than ldapo?

Sidenote: Reading the code, is it true that ldapo does everything what the bash scripts do, except ``clear`` ? (If yes, why  not to have ``clear`` included too ?)

Would suggest ``ldapmanager``

joshmoore committed:
ldapo: add `clear` command

Matches bin/clear but additionally allows choosing
the distinguished name (dn) for the *subtree* to
delete.

joshmoore commented:
* `clear` added
 * `ldapmanager` makes sense to me. Do we want any thing ome-like in the name though? (Additionally, should this go to PyPI?)
 * Another question I had: how annoying is the `--apply` flag? Should it be the default?
 * What is the suggested future of the other scripts?
 * Any other thoughts?

joshmoore committed:
ldapo: add default argument for search

This makes searching much more friendly since
by default all objects will be printed.

pwalczysko commented:
>Do we want any thing ome-like in the name though? (Additionally, should this go to PyPI?)

Maybe it could go to PyPI. But, except that this is work of an OME Team member, what is OME on it ? I understand that this is rather supplementing the user-friendliness of LDAP ? But if we want to have any ``ome`` in the name how about ``ldapomenager`` ?

pwalczysko commented:
>... how annoying is the --apply flag? Should it be the default?

the flag is confusing, I would go for that to be default

pwalczysko commented:
>What is the suggested future of the other scripts?

Supposing that all the functionality is already in the new ``ldapomenager`` , then hmmmm. Maybe we could have them in reserve in case somebody cannot install python-ldap ?

pwalczysko commented:
>Any other thoughts?

No.

joshmoore commented:
> what is OME on it ?

The hard-coded DN: `dc=openmicroscopy,dc=org`. This is what I'd call an opinionated tool in that it assumes:
```
openmicroscopy.org/
├── groups
│   ├── g1
│   └── g2
└── users
    ├── u1
    └── u2
```

> the flag is confusing, I would go for that to be default

ok. Then I will switch `--apply` to `--dry-run` and reverse the logic

> Maybe we could have them in reserve in case somebody cannot install python-ldap ?

No objections there. That's also why `ldapo` prints out commands on dry-run. These can be passed to `ldapadd` and `ldapmodify`

joshmoore committed:
ldapo: --apply is now default; --dry-run exists

joshmoore committed:
ldapo: rename to **ldapmanager**

joshmoore commented:
Pushed.

sbesson commented:
Re `ldapmanager`, agreed on the OME assumptions. Maybe this decision depends on the scope in terms of deployment/publication. If this script remains something internal i.e.  someone would need to `wget` it with the hardcoded URL, it might be fine as it is for now.
If we deploy it on PyPI or similar so that the manager utility gets installed alongside the classical OpenLDAP utilities or other third-party LDAP utilities, it might be useful to differentiate them.

joshmoore commented:
Ok. I'm going to assume we're *not* pypi-ing at this stage, and therefore no further changes are needed (if not requested by @pwalczysko).

joshmoore commented:
Thanks guys. Releasing as 0.3.0 then.
